File: src/lib/planNames.js
/*
  Ein-Wort-Namen für Tagespläne.

  KI-Pläne bekommen ihren Namen vom Modell (es kennt den Inhalt). Gewürfelte
  Pläne haben kein Modell im Spiel und würden sonst namenlos bleiben — die
  holen sich hier einen. Beides ist bewusst ein einzelnes Wort: es soll ein
  Rufname sein ("Titan am Mittwoch"), keine Beschreibung.
*/

// Wörter, die in beiden Sprachen funktionieren und nach Kraft klingen.
const NAMES = [
  "Titan", "Granit", "Vulkan", "Anker", "Kobalt", "Falke", "Orkan", "Basalt",
  "Zenit", "Nova", "Atlas", "Krater", "Quarz", "Bison", "Komet", "Obsidian",
  "Kanon", "Wolf", "Magma", "Bastion", "Kobra", "Zunder", "Achat", "Delta",
  "Fenrir", "Gletscher", "Hammer", "Impuls", "Jaguar", "Kaskade",
];

/*
  Gleicher Plan -> gleicher Name, auch nach einem Reload. Deshalb wird aus dem
  Seed ein Hash gebildet statt zufällig zu ziehen; sonst hieße der Mittwoch bei
  jedem Rendern anders.
*/
export function planNameFor(seed) {
  const text = String(seed ?? "");
  let hash = 0;
  for (let i = 0; i < text.length; i++) {
    hash = (hash * 31 + text.charCodeAt(i)) | 0;
  }
  return NAMES[Math.abs(hash) % NAMES.length];
}

/*
  Ein Name, der noch nicht vergeben ist.

  "Nicht wiederholen" muss über alles hinweg gelten, was der Nutzer sieht —
  gespeicherte Pläne UND Favoriten. Deshalb kommt die Liste der belegten Namen
  von außen, statt hier geraten zu werden. Sind alle Wörter verbraucht, wird
  durchnummeriert (Titan 2), damit nie ein Duplikat entsteht.
*/
export function uniquePlanName(taken = [], seed = "") {
  const used = new Set([...taken].filter(Boolean).map((n) => String(n).toLowerCase()));

  const start = Math.abs(hash(String(seed))) % NAMES.length;
  for (let i = 0; i < NAMES.length; i++) {
    const candidate = NAMES[(start + i) % NAMES.length];
    if (!used.has(candidate.toLowerCase())) return candidate;
  }

  for (let round = 2; round < 100; round++) {
    for (const base of NAMES) {
      const candidate = `${base} ${round}`;
      if (!used.has(candidate.toLowerCase())) return candidate;
    }
  }
  return `Plan ${used.size + 1}`;
}

function hash(text) {
  let h = 0;
  for (let i = 0; i < text.length; i++) h = (h * 31 + text.charCodeAt(i)) | 0;
  return h;
}

/* Ein Name pro Tag, ohne Dopplungen innerhalb desselben Plans. */
export function planNamesForDays(days, salt = "") {
  const taken = new Set();
  const out = {};
  days.forEach((day) => {
    let name = planNameFor(`${salt}:${day}`);
    let bump = 0;
    while (taken.has(name) && bump < NAMES.length) {
      bump++;
      name = planNameFor(`${salt}:${day}:${bump}`);
    }
    taken.add(name);
    out[day] = name;
  });
  return out;
}

File: Kraftwuerfel/Models/PlanNames.swift
import Foundation

/*
  Portierung von src/lib/planNames.js.

  Gleicher Plan -> gleicher Name, auch nach einem Neustart: der Name kommt aus
  einem Hash des Seeds, nicht aus dem Zufall. Der Hash rechnet wie in JavaScript
  mit 32-Bit-Überlauf (`| 0`) und über UTF-16-Einheiten, sonst vergäbe die App
  andere Namen als das Web für denselben Plan.
*/
public enum PlanNames {

    public static let names = [
        "Titan", "Granit", "Vulkan", "Anker", "Kobalt", "Falke", "Orkan", "Basalt",
        "Zenit", "Nova", "Atlas", "Krater", "Quarz", "Bison", "Komet", "Obsidian",
        "Kanon", "Wolf", "Magma", "Bastion", "Kobra", "Zunder", "Achat", "Delta",
        "Fenrir", "Gletscher", "Hammer", "Impuls", "Jaguar", "Kaskade",
    ]

    static func hash(_ text: String) -> Int32 {
        var h: Int32 = 0
        for unit in text.utf16 {
            h = h &* 31 &+ Int32(unit)
        }
        return h
    }

    private static func index(_ seed: String) -> Int {
        let h = hash(seed)
        return Int(h.magnitude % UInt32(names.count))
    }

    public static func planName(for seed: String) -> String {
        names[index(seed)]
    }

    /// Ein Name, der weder unter den gespeicherten Plänen noch unter den
    /// Favoriten schon vergeben ist. Sind alle Wörter belegt, wird nummeriert.
    public static func uniquePlanName(taken: [String] = [], seed: String = "") -> String {
        let used = Set(taken.filter { !$0.isEmpty }.map { $0.lowercased() })

        let start = index(seed)
        for i in 0..<names.count {
            let candidate = names[(start + i) % names.count]
            if !used.contains(candidate.lowercased()) { return candidate }
        }

        for round in 2..<100 {
            for base in names {
                let candidate = "\(base) \(round)"
                if !used.contains(candidate.lowercased()) { return candidate }
            }
        }
        return "Plan \(used.count + 1)"
    }

    /// Ein Name pro Tag, ohne Dopplungen innerhalb desselben Plans.
    public static func planNamesForDays(_ days: [String], salt: String = "") -> [String: String] {
        var taken = Set<String>()
        var out: [String: String] = [:]
        for day in days {
            var name = planName(for: "\(salt):\(day)")
            var bump = 0
            while taken.contains(name) && bump < names.count {
                bump += 1
                name = planName(for: "\(salt):\(day):\(bump)")
            }
            taken.insert(name)
            out[day] = name
        }
        return out
    }
}
